provider/docker_test.go

```diff
@@ -2271,3 +2275,93 @@ func TestSwarmTaskParsing(t *testing.T) {
 		}
 	}
 }
+
+type fakeTasksClient struct {
+	dockerClient.APIClient
+	tasks []swarm.Task
+	err   error
+}
+
+func (c *fakeTasksClient) TaskList(ctx context.Context, options dockertypes.TaskListOptions) ([]swarm.Task, error) {
+	return c.tasks, c.err
+}
+
+func TestListTasks(t *testing.T) {
+	cases := []struct {
+		service       swarm.Service
+		tasks         []swarm.Task
+		isGlobalSvc   bool
+		expectedTasks []string
+		networks      map[string]*docker.NetworkResource
+	}{
+		{
+			service: swarm.Service{
+				Spec: swarm.ServiceSpec{
+					Annotations: swarm.Annotations{
+						Name: "container",
+					},
+				},
+			},
+			tasks: []swarm.Task{
+				{
+					ID:   "id1",
+					Slot: 1,
+					Status: swarm.TaskStatus{
+						State: swarm.TaskStateRunning,
+					},
+				},
+				{
+					ID:   "id2",
+					Slot: 2,
+					Status: swarm.TaskStatus{
+						State: swarm.TaskStatePending,
+					},
+				},
+				{
+					ID:   "id3",
+					Slot: 3,
+				},
+				{
+					ID:   "id4",
+					Slot: 4,
+					Status: swarm.TaskStatus{
+						State: swarm.TaskStateRunning,
+					},
+				},
+				{
+					ID:   "id5",
+					Slot: 5,
+					Status: swarm.TaskStatus{
+						State: swarm.TaskStateFailed,
+					},
+				},
+			},
+			isGlobalSvc: false,
+			expectedTasks: []string{
+				"container.1",
+				"container.4",
+			},
+			networks: map[string]*docker.NetworkResource{
+				"1": {
+					Name: "foo",
+				},
+			},
+		},
+	}
+
+	for _, e := range cases {
+		dockerData := parseService(e.service, e.networks)
```

Comment:
I'd really like us to use more [sub-tests](https://blog.golang.org/subtests). Benefits:

- We can run sub-tests individually (`go test -run <sub test specifier>`)
- We are encouraged to pass a meaningful context into the `t.Run()` method (and, at the same time, don't need to add it at every `t.{Log,Error,Fatal}*` call).
- We can parallelize tests easily.
- If we use `t.Fatal*`, we only fatal out of the current sub-test and not the entire one.

Setting up sub-tests are really easy: [Example](https://github.com/timoreimann/traefik/blob/aae80ab324974cdb58f4cbed103caa15b0b3ba5e/provider/kubernetes_test.go#L328-L331).